sound/soc/intel/boards/skl_hda_dsp_generic.c

```diff
@@ -168,6 +168,26 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
 	return 0;
 }
 
+#define HDA_ANA_CODEC_AUTOSUSPEND_DELAY_MS 2000
+
+static void skl_hda_set_ana_codec_autosuspend(struct snd_soc_card *card)
+{
+	struct snd_soc_dai *codec_dai;
+	struct hdac_hda_priv *hda_pvt;
+	struct hdac_device *hdev;
+
+	codec_dai = snd_soc_card_get_codec_dai(card, "Analog Codec DAI");
+	if (!codec_dai)
+		return;
+
+	hda_pvt = snd_soc_component_get_drvdata(codec_dai->component);
+	hdev = &hda_pvt->codec.core;
+
+	pm_runtime_set_autosuspend_delay(&hdev->dev,
+					 HDA_ANA_CODEC_AUTOSUSPEND_DELAY_MS);
+	pm_runtime_use_autosuspend(&hdev->dev);
+}
```

Comment:
I am not clear on the following statement:

"And the analog codec needs to enter runtime_suspend at least once,
otherwise the codec can't output the sound normally. So we set the
autosuspend after the snd_card is registered, this will not prevent
the codec from entering runtime_suspend during registering snd_card."

what makes you think the card registration will only complete *after* the codec enters pm_runtime_suspend?

This looks racy to me.

Reply:
@plbossart Agreed. @jason77-wang @juimonen I think it's good we now have a patch that fixes both problems (in one patch), but we still need to polish the fix to analog codec issue, before this is upstream ready. 

If power cycle is needed, that sounds like something in the codec init sequence is not correct. Perhaps a forced power cycle could be implemented on the codec side as well?

Reply:
So again I repeat my point, which I've said now about 20 times in all possible related PR's. We should in anyway set the autosuspend delay, even if it produces some side effect in some platform. It is clear abnormal situation that it is not set at all currently...

And agreed, that if some particular codec needs "artificial" up/down sequences in boot, we should not really try to fix that in same patch. It is good if this kind of patch is for some reason found, but I don't think it makes sense to try to find this kind of spot in the code. We are then essentially saying: "normal/standard sof bootup is that you need to power up/down the codec first before setting any autosuspend delay".

Reply:
Today a couple of new LENOVO models arrived our office, they all have this issue (mic can't be detected), and even applying my patch, the mic still can't be detected. Looks like only setting autosuspend_delay is not enough to fix the problem for those new machines.

Reply:
@jason77-wang did you try to disable the pm altogether, so setting the suspend to 0? Or making the suspend very long like 5-10s? I recall talking with realtek folks that for .e.g. the windows driver has much longer suspend delay...

Reply:
If I disable the runtime_pm from the codec, the mic could be detected. (since slave is a slave of hda controller, that means both codec and controller's runtime pm are disabled).

If I set the delay to 5~10, it doesn't help, the mic can't be detected.

If I set the delay to 1s and disable controller's runtime_pm (snd-sof-pci.sof-pci-debug=0x1), the mic could be detected. (that means only enable the runtime_pm on codec, disable the runtime_pm on controller & dsp)

So to fix this problem, it looks like we need to do 2 changes: add a delay to codec's rt_pm, refine the controller's rt_pm (suspend and resume).

Reply:
@jason77-wang Thanks, this is good analysis. Can you test how the legacy driver behaves on these machines? 

There are now many issues grouped together, so we need to get moving and solve issues piece by piece. 

One proposal would be to implement a mechanism like sound/pci/hda/hda_intel.c:power_save_blacklist to SOF skl_hda_dsp_generic.c and blacklist the small set of devices with codecs that do not work correct with runtime PM. This would then allow us to proceed and merge a patch that sets autosuspend delay to a non-zero value, and align us with behaviour of the legacy driver (i.e. merge #1726).

@juimonen @jason77-wang ?
Any takers for the first patch (blacklisting Mantis and the new Lenovo devices)?